src/ripple/unl/tests/Sim3.h

```diff
@@ -0,0 +1,513 @@
+            std::shared_ptr<PosMsg const> const& m)
+        {
+            if (round_->receive(*m))
+                relay(from, m);
+            else
+                ++net_.dup;
+        }
+
+        void
+        timer()
+        {
+            if (round_->update(unl_, net_.now()))
+                broadcast(round_->posMsg());
+            if (round_->consensus_)
+                return;
+            using namespace std::chrono;
+            net_.timer(milliseconds(nUpdateMS),
+                [=]() { timer(); });
+        }
+
+        //----------------------------------------------------------------------
+
+        // Send a message to this peer
+        template <class Message>
+        void
+        send (Peer& from,
+            std::shared_ptr<Message const> const& m)
+        {
+            ++net_.sent;
+            net_.send (from, *this,
+                [&, m]() { receive(from, m); });
+        }
+
+        // Broadcast a message to all links
+        template <class Message>
+        void
+        broadcast (std::shared_ptr<
+            Message const> const& m)
+        {
+            for(auto& link : net_.links(*this))
+                link.to.send(*this, m);
+        }
+
+        // Relay a message to all links
+        template <class Message>
+        void
+        relay (Peer& from,
+            std::shared_ptr<Message const> const& m)
+        {
+            for(auto& link : net_.links(*this))
+                if (&link.to != &from)
+                    link.to.send(*this, m);
+        }
+    };
+
+    //--------------------------------------------------------------------------
+
+    // The result of one round
+    struct Result
+    {
+        std::size_t elapsed;
+        std::size_t failure = 0;
+        std::size_t consensus = 0;
+        std::set<ItemSet> sets;
+    };
+
+    // The results of several rounds
+    struct Results
+    {
+        std::size_t rounds = 0;
+        std::size_t perfect = 0;
+        std::vector<std::size_t> elapsed;
+        std::vector<std::size_t> failure;
+        std::vector<std::size_t> consensus;
+
+        void
+        aggregate (Result const& result)
+        {
+            ++rounds;
+            perfect += result.sets.size() == 1;
+            elapsed.push_back(result.elapsed);
+            failure.push_back(result.failure);
+            consensus.push_back(result.consensus);
+        }
+    };
+
+    struct Report
+    {
+        std::size_t perfect;
+        std::size_t elapsed_min;
+        std::size_t elapsed_max;
+
+        Report (Results& results, Config const& config)
+        {
+            perfect = results.perfect;
+#if 0
+            std::sort(
+                results.elapsed.begin(), results.elapsed.end());
+            std::sort(
+                results.consensus.begin(), results.consensus.end(),
+                    std::greater<std::size_t>{});
+#endif
+            elapsed_min = results.elapsed.front();
+            elapsed_max = results.elapsed.back();
```

Comment:
It looks like `elapsed` is inserted sorted, but given the `sort` above `std::minmax_element` would work as well.